### ISSUE TITLE
Issue with path.repo while starting Elasticsearch as a service

### ISSUE THREAD

**Issue by viswanadh9** [elastic/elasticsearch#50117]
**Describe the feature**: Elasticsearch service doesn't get started if path.repo contains a repository path that needs credentials. 

**Elasticsearch version** (`bin/elasticsearch --version`): 6.5.4

**Plugins installed**: [Search guard]

**JVM version** (`java -version`): 1.8

**OS version** (`uname -a` if on a Unix-like system): Windows 10, Windows Server 2012 R2

**Description of the problem including expected versus actual behavior**:

When we give a repository path which needs credentials in **path.repo** & we connect to that repository using [net use](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2012-r2-and-2012/gg651155(v%3Dws.11)) command, elasticsearch doesn't start as a service. However, elasticsearch.bat works perfectly fine.

Fails with error message like this: "Exception java.lang.IllegalStateException: Unable to access 'path.repo', Caused by: java.nio.file.FileSystemException, **Unable to determine if root directory exists**"

**Steps to reproduce**:

 1. Open elasticsearch.yml & add a repository path to path.repo. The repository path should be asking for credentials (may be give cross domain machine's share)
 2. Now execute this command to connect to the share: "net use '\\machine-name\share1'"
 3. Once you are connected, start Elasticsearch as a service. Fails with above mentioned exception. Try starting the same using "elasticsearch.bat", it works perfectly. 

**Provide logs (if relevant)**:
[Elasticsearch Logs.zip](https://github.com/elastic/elasticsearch/files/3954174/Elasticsearch.Logs.zip)

**Comment by rjernst** [comment]
Do you attempt to launch the Elasticsearch service immediately after running the `net use` command? Are you sure there is no delay in the share being available? Can you try running elasticsearch.bat in place of the Elasticsearch service and see if it fails?

**Comment by viswanadh9** [comment]
Thanks for the quick response. Yeah, I am very sure the share is available at the moment I started Elasticsearch service (Tried opening that share in Windows Explorer & was able to read/write data from/to that share). As I already mentioned, **this works perfectly fine with elasticsearch.bat & not with Elasticsearch service**. I really could not get the difference b/w this elasticsearch.bat & elasticsearch service.

**Note:** I have also tried this with latest version of Elasticsearch i.e., 7.5, same issue persists. 

**Addl Info:** From the stack trace inside CLUSTER logs, I found out that this API (**java.nio.file.Files.createDirectories**) throws the reported error message "**Unable to determine if root directory exists**". I have even tried invoking this API in a standalone Java in the same machine, it worked fine. But I can't get it to work in this Elasticsearch service.

**Comment by rjernst** [comment]
> Tried opening that share in Windows Explorer & was able to read/write data from/to that share

This doesn't guarantee the share is available immediately, only after the time it takes for you to run the `net use` and then open windows explorer.

> this works perfectly fine with elasticsearch.bat 

Again, I just want to confirm, when you test with elasticsearch.bat, are you doing this *in place* of where you would launch the elasticsearch service (in whatever script is running `net use`)? Or are you manually starting it after the elasticsearch service fails?

**Comment by viswanadh9** [comment]
> This doesn't guarantee the share is available immediately, only after the time it takes for you to run the net use and then open windows explorer.

I am pretty sure the share is available. Before executing net use command, it was not accessible. After execution of this (`net use`) command, I was even able to use the API (where Elasticsearch service fails) - **java.nio.file.Files.createDirectories** successfully. I suppose availability of share is nothing but we being able to see it, read/write data. I've even waited for so long to confirm its availability. Had the share been unavailable, why would `elasticsearch.bat` work fine !!

> Again, I just want to confirm, when you test with elasticsearch.bat, are you doing this in place of where you would launch the elasticsearch service (in whatever script is running net use)? Or are you manually starting it after the elasticsearch service fails?

I run everything separately in command prompts & am not using any kind of scripts (or combination of commands). Here's what I did:
1. Open Command prompt 1 (as administrator) & execute `net use` command.
2. Open Command prompt 2 (as administrator) & execute `elasticsearch.bat`. This works fine. Stop this bat.
3. Go to `services.msc`, find Elasticsearch service & start it. **Fails with reported error message.**

We can do a remote session if required. I am ready for that too.

**Comment by rjernst** [comment]
> Go to services.msc

Can you please try starting it with `elasticsearch-service.bat start`? Can you also please confirm you installed the service with `elasticsearch-service.bat install`?

**Comment by viswanadh9** [comment]
> Can you please try starting it with `elasticsearch-service.bat start`?

I tried this too, even this din't work.

> Can you also please confirm you installed the service with `elasticsearch-service.bat install`?

Yes I did install using this command.

**Comment by viswanadh9** [comment]
Hi @rjernst, Do we have any update on this? I can send you any further details if required, please investigate on this asap.

**Comment by rjernst** [comment]
@viswanadh9 If you are in need of a service with an SLA that covers response times for questions then you may want to consider talking to us about a [subscription](https://www.elastic.co/subscriptions).

I have been unable to reproduce this, and it sounds like an environmental issue. We use github for verified bug reports, feature requests, and pull requests. There's an active community in the [forums](https://discuss.elastic.co) that should be able to help get an answer to your question. As such, I hope you don't mind that I close this.

**Comment by rjernst** [comment]
Also, one more thing for your to consider in your debugging. Ensure the user the service runs as has access to the share. If the service user is different and doesn't have access this might explain why when you run elasticsearch.bat the share is readable.

**Comment by viswanadh9** [comment]
> I have been unable to reproduce this, and it sounds like an environmental issue.

I really can't believe that you were not able to reproduce this issue. I tried not only in my personal machine, we do have a lab of Virtual Machines and tried in all sort of machines (ones in a domain, ones in work group). 

> There's an active community in the forums that should be able to help get an answer to your question

I have already done this on the same day I raised that issue here. Still there's no reply yet. I lost my hope on that. Thanks for the suggestion :)

> Also, one more thing for your to consider in your debugging. Ensure the user the service runs as has access to the share. If the service user is different and doesn't have access this might explain why when you run elasticsearch.bat the share is readable.

Yeah, I considered this. Tried all possible ways to by changing the accounts in `Logon` tab. Din't work.

**Comment by viswanadh9** [comment]
**Final Update (and Solution):** Am posting 'why this issue occurred & how I fixed it' here so that it will be helpful for others.

As I already said, if we gave a network share (**which requires credentials**, for instance assume cross-domain share) in `path.repo` of elasticsearch.yml, elasticsearch service doesn't start even if we are connected to it using `net use` command. 

This happened because service runs entirely in a different context i.e., it runs using System account (ntauthority/system) isolated for the currently logged on user (More info [here](https://serverfault.com/questions/177139/windows-service-cant-access-network-share)). 

So we need to run the `net use` command in the context of service (or system account), so I followed this link [here](https://davidhamann.de/2018/08/17/running-script-as-local-system-windows/) to do this. And finally the elasticsearch service got started. 

**Details Soln:** 
1. Download PSTools
2. Execute `psexec -i -s cmd.exe`
3. Now execute `net use` command in the command prompt opened. Try starting the service, it works.
